libs/nio/src/main/java/org/elasticsearch/nio/NioSelector.java

```diff
@@ -238,12 +238,13 @@ void processKey(SelectionKey selectionKey) {
             }
 
             if (channelContext.isConnectComplete()) {
-                if ((ops & SelectionKey.OP_WRITE) != 0) {
-                    handleWrite(channelContext);
-                }
-
-                if ((ops & SelectionKey.OP_READ) != 0) {
-                    handleRead(channelContext);
+                if (channelContext.selectorShouldClose() == false) {
+                    if ((ops & SelectionKey.OP_WRITE) != 0) {
+                        handleWrite(channelContext);
+                    }
+                    if (channelContext.selectorShouldClose() == false && (ops & SelectionKey.OP_READ) != 0) {
```

Comment:
We can run into this here if the write ready flag event was set to signal input shutdown on the other side and shouldn't try to read here since it will fail anyway.